src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java

```diff
@@ -117,94 +125,148 @@ public String getScript() {
+                        controlDir.getRemote(),
+                        launcherAgent.getRemote(),
+                        scriptPath,
+                        cookieValue,
+                        cookieVariable);
+            } else {
+                launcherCmd = scriptLauncherCmd(c, ws, shell, os, scriptPath, cookieValue, cookieVariable);
+            }
         }
-        if (os == OsType.WINDOWS) { // JENKINS-40255
-            scriptPath= scriptPath.replace("\\", "/"); // cygwin sh understands mixed path  (ie : "c:/jenkins/workspace/script.sh" )
+
+        LOGGER.log(Level.FINE, "launching {0}", launcherCmd);
+        Launcher.ProcStarter ps = launcher.launch().cmds(launcherCmd).envs(escape(envVars)).pwd(ws).quiet(true);
+        if (LAUNCH_DIAGNOSTICS) {
+            ps.stdout(listener);
+        } else {
+            ps.readStdout().readStderr(); // TODO RemoteLauncher.launch fails to check ps.stdout == NULL_OUTPUT_STREAM, so it creates a useless thread even if you never called stdout(…)
         }
+        ps.start();
+        return c;
+    }
 
-        envVars.put(cookieVariable, "please-do-not-kill-me");
-        // The temporary variable is to ensure JENKINS_SERVER_COOKIE=durable-… does not appear even in argv[], lest it be confused with the environment.
-        String cmd;
+    private List<String> binaryLauncherCmd(ShellController c, FilePath ws, String shell, String controlDirPath, String binaryPath, String scriptPath, String cookieValue, String cookieVariable) throws IOException, InterruptedException {
+        String logFile = c.getLogFile(ws).getRemote();
+        String resultFile = c.getResultFile(ws).getRemote();
+        String outputFile = c.getOutputFile(ws).getRemote();
+
+        List<String> cmd = new ArrayList<>();
+        cmd.add(binaryPath);
+        cmd.add("-controldir=" + controlDirPath);
+        cmd.add("-result=" + resultFile);
+        cmd.add("-log=" + logFile);
+        cmd.add("-cookiename=" + cookieVariable);
+        cmd.add("-cookieval=" + cookieValue);
+        cmd.add("-script=" + scriptPath);
+        if (shell != null) {
+            cmd.add("-shell=" + shell);
+        }
+        if (capturingOutput) {
+            cmd.add("-output=" + outputFile);
+        }
+        if (!LAUNCH_DIAGNOSTICS) {
+            // JENKINS-58290: launch in the background. No need to close stdout/err, binary does not write to them.
+            cmd.add("-daemon");
+        }
+        return cmd;
+    }
+
+    private List<String> scriptLauncherCmd(ShellController c, FilePath ws, String shell, OsType os, String scriptPath, String cookieValue, String cookieVariable) throws IOException, InterruptedException {
+        String cmdString;
         FilePath logFile = c.getLogFile(ws);
         FilePath resultFile = c.getResultFile(ws);
         FilePath controlDir = c.controlDir(ws);
+        String interpreter = "";
+
+        if (!script.startsWith("#!")) {
+            interpreter = "'" + shell + "' -xe ";
```

Comment:
It's not totally obvious that `shell` is guaranteed non-null in this case. Might be good to add nullability annotations to all of the parameters of this method and `binaryLauncherCmd` so Spotbugs can help make sure we don't introduce any NPEs for untested combinations for untested combinations of values later on.

Reply:
You added `@Nonnull` to shell, but IIUC it _is_ null if `!script.startsWith("#!")`, since it gets initialized to `null` in `launchWithCookie` and is only set if `script.startsWith("#!")`. Is that not the case?

Reply:
ugh, you're totally right. I got too paste happy. I even remembering telling myself not to do that :(

Reply:
nit: Kinda done in 5f7d04ed5, but I think it would be better to use `@Nullable` or `@CheckForNull` rather than just not annotating the things that aren't `@Nonnull`. You'd have to change conditions like the `if (!script.startsWith("#!")) { ... some code that assumes shell is nonnull ... }` to `if (!script.startsWith("#!") && shell != null) { ... }` to make SpotBugs happy, but that's probably clearer anyway.

Reply:
that makes way more sense

Reply:
oddly enough, SpotBugs didn't complain when I did not add the `shell != null` for `@CheckForNull`. I still kept it in there though. For `binaryLauncherCmd` i just had it as `@Nullable`